engine/src/main/java/org/hibernate/validator/internal/metadata/aggregated/AbstractConstraintMetaData.java

```diff
@@ -168,79 +166,56 @@ else if ( !name.equals( other.name ) ) {
+					asDescriptors( childContainerElementMetaDataTree.constraints ),
+					childrenDescriptors,
+					childContainerElementMetaDataTree.cascading,
+					defaultGroupSequenceRedefined, defaultGroupSequence,
+					childContainerElementMetaDataTree.groupConversionDescriptors ) );
+		}
 
-			Type[] typeArguments = ( (ParameterizedType) type ).getActualTypeArguments();
-			TypeVariable<?>[] typeParameters = ReflectionHelper.getClassFromType( type ).getTypeParameters();
+		return containerElementTypeDescriptors;
+	}
 
-			for ( int i = 0; i < typeArguments.length; i++ ) {
-				Type typeArgument = typeArguments[i];
-				TypeVariable<?> typeParameter = typeParameters[i];
+	private static class ContainerElementMetaDataTree {
 
-				Set<MetaConstraint<?>> constraints = Collections.emptySet();
-				ContainerElementMetaConstraintTree currentContainerElementMetaConstraintTree = null;
+		private Map<TypeVariable<?>, ContainerElementMetaDataTree> nodes = new HashMap<>();
 
-				if ( containerElementMetaConstraintTree != null && containerElementMetaConstraintTree.nodes.containsKey( typeParameter ) ) {
-					currentContainerElementMetaConstraintTree = containerElementMetaConstraintTree.nodes.get( typeParameter );
-					constraints = containerElementMetaConstraintTree.nodes.get( typeParameter ).constraints;
-				}
+		private Type elementType = null;
 
-				CascadingMetaData currentCascadingMetaData = null;
-				boolean cascading = false;
-				Set<GroupConversionDescriptor> groupConversionDescriptors = Collections.emptySet();
-
-				if ( cascadingMetaData != null ) {
-					for ( CascadingMetaData candidateCascadingMetaData : cascadingMetaData.getContainerElementTypesCascadingMetaData() ) {
-						if ( candidateCascadingMetaData.getTypeParameter().equals( typeParameter ) ) {
-							currentCascadingMetaData = candidateCascadingMetaData;
-							cascading = currentCascadingMetaData.isCascading();
-							groupConversionDescriptors = currentCascadingMetaData.getGroupConversionDescriptors();
-						}
-					}
-				}
+		private Class<?> containerClass;
 
-				containerElementTypeDescriptors.add( new ContainerElementTypeDescriptorImpl(
-						typeArgument, TypeVariables.getTypeParameterIndex( typeParameter ),
-						asDescriptors( constraints ),
-						asContainerElementTypeDescriptors( typeArgument, currentContainerElementMetaConstraintTree, currentCascadingMetaData,
-								defaultGroupSequenceRedefined, defaultGroupSequence ),
-						cascading,
-						defaultGroupSequenceRedefined, defaultGroupSequence,
-						groupConversionDescriptors ) );
-			}
-
-			return containerElementTypeDescriptors;
-		}
-		else if ( type instanceof GenericArrayType ) {
-			// TODO Container element constraints are not supported for arrays in the Bean Validation specification.
-			// Let's ignore them for now.
-			return Collections.emptyList();
-		}
-		else {
-			return Collections.emptyList();
-		}
-	}
-
-	private static class ContainerElementMetaConstraintTree {
```

Comment:
Is this type only short-living during initialisation or are instances kept at runtime? In case of the latter I'm a bit concerned about the non-final fields in here. A short class comment would be nice, too.